service/storage_service.cc

```diff
@@ -171,19 +171,10 @@ void storage_service::register_metrics() {
 }
 
 std::optional<gms::inet_address> storage_service::get_replace_address() {
-    auto& cfg = _db.local().get_config();
-    sstring replace_address = cfg.replace_address();
-    sstring replace_address_first_boot = cfg.replace_address_first_boot();
-    try {
-        if (!replace_address.empty()) {
-            return gms::inet_address(replace_address);
-        } else if (!replace_address_first_boot.empty()) {
-            return gms::inet_address(replace_address_first_boot);
-        }
-        return std::nullopt;
-    } catch (...) {
-        return std::nullopt;
+    if (!_replacement_info) {
+        on_internal_error(slogger, "replacement_info is unset");
```

Comment:
You could quickly check it in your test.py tests by (temporarily just to see, I don't suggest to send a patch with this) including `'enable_repair_based_node_ops': False` in `make_scylla_conf` in pylib/scylla_cluster.py

Reply:
Still, main -> join_cluster -> join_token_ring must complete before on_change -> handle_state_replacing happens.
The event-based state machine relies on gossip to transition the replacing node between those states.
Other nodes will process the transition of the replacing node as part of on_change -> join_cluster.

Is that right @asias?

FWIW, test_topology passes with `'enable_repair_based_node_ops': False in make_scylla_conf`

Reply:
with/without enable_repair_based_node_ops == true  for replace, handle_state_replacing is not going to be called anymore, since the replacing node will not publish the HIBERNATE state any more.

Reply:
A node with new code running replace will not publish HIBERNATE state, so handle_state_replacing on new and old node will not be called. 

But if an old node runs replacing, it will publish HIBERNATE, so handle_state_replacing will be called. In this case, the replacing node will use the same host_id as the node to be replaced anyway. So even if in your make host id unique patch, it does not handle the case replacing node uses new host id in handle_state_replacing , it should be fine.

Reply:
> Still, main -> join_cluster -> join_token_ring must complete before on_change -> handle_state_replacing happens.

handle_state_replacing will also happen on other nodes, not only the one which is replacing. These nodes will not have `_replacement_info`.

> But if an old node runs replacing, it will publish HIBERNATE, so handle_state_replacing will be called.

Let's assume that we don't run replace in the middle of upgrade, so in particular we don't run replace with "old nodes".

But that doesn't change anything, the user can still use `'enable_repair_based_node_ops': False` in a **new node**, and that node will publish HIBERNATE when replacing, which will cause other nodes to call `handle_state_replacing`, right?

Reply:
If `'enable_repair_based_node_ops': False` is a legal thing to do, we must continue supporting it. Only once we completely get rid of the option, we can assume that nodes won't publish HIBERNATE and so on.

Reply:
I don't understand how does the current code supposed work with get_replace_address.
It would return a disengaged optional on nodes other than the replacing one.
We can support that

Reply:
https://github.com/scylladb/scylladb/issues/12330

Reply:
> It would return a disengaged optional on nodes other than the replacing one.

Yes.

But after understanding that `handle_state_replacing` should no longer be ever called, I'm proposing to remove it (outside of this PR), and then we'll easily see that `get_replace_address` is called only by the replacing node.

For now, I'm okay with the runtime check - `handle_state_replacing` can be broken since it's dead code anyway.

Reply:
works for me